File: PinetreeShop/PinetreeShop.Domain.Products/ProductAggregate.cs
using System;
using PinetreeCQRS.Infrastructure;
using PinetreeShop.Domain.Products.Commands;
using PinetreeShop.Domain.Products.Events;
using PinetreeShop.Domain.Products.Exceptions;

namespace PinetreeShop.Domain.Products
{
    public class ProductAggregate : AggregateBase
    {
        public string Name { get; set; }
        public decimal Price { get; set; }
        public int Quantity { get; set; }
        public int Reserved { get; set; }
        private int AvailableQuantity { get { return Quantity - Reserved; } }

        #region Constructors

        public ProductAggregate()
        {
            Quantity = 0;
            Reserved = 0;

            RegisterEventHandler<ProductCreated>(Apply);
            RegisterEventHandler<ProductQuantityChanged>(Apply);
            RegisterEventHandler<ProductReserved>(Apply);
            RegisterEventHandler<ProductReservationCancelled>(Apply);
            RegisterEventHandler<ReservedProductPurchased>(Apply);
        }

        private ProductAggregate(CreateProduct cmd) : this()
        {
            RaiseEvent(new ProductCreated(cmd.AggregateId, cmd.Name, cmd.Price));
        }

        #endregion

        #region Event handlers

        private void Apply(ReservedProductPurchased obj)
        {
            Quantity -= obj.Quantity;
            Reserved -= obj.Quantity;
        }

        private void Apply(ProductReservationCancelled evt)
        {
            Reserved -= evt.Quantity;
        }

        private void Apply(ProductReserved evt)
        {
            Reserved += evt.QuantityToReserve;
        }

        private void Apply(ProductQuantityChanged evt)
        {
            Quantity += evt.Difference;
        }

        private void Apply(ProductCreated evt)
        {
            AggregateId = evt.AggregateId;
            Name = evt.Name;
            Price = evt.Price;
        }

        #endregion

        #region Command handlers

        internal static ProductAggregate Create(CreateProduct cmd)
        {
            if (cmd.Price <= 0) throw new ProductCreationException(cmd.AggregateId, $"Price {cmd.Price} must be a positive value.");

            return new ProductAggregate(cmd);
        }

        internal void AddToStock(AddProductToStock cmd)
        {
            var newQuantity = Quantity + cmd.Quantity;
            ChangeQuantity(newQuantity);
        }

        internal void PurchaseReserved(PurchaseReservedProduct cmd)
        {
            // TODO: throw exception if reserved or available less than quantity

            RaiseEvent(new ReservedProductPurchased(cmd.AggregateId, cmd.Quantity));
        }

        internal void RemoveFromStock(RemoveProductFromStock cmd)
        {
            var newQuantity = Quantity - cmd.Quantity;
            ChangeQuantity(newQuantity);
        }

        internal void SetQuantity(SetProductQuantity cmd)
        {
            ChangeQuantity(cmd.Quantity);
        }

        internal void CancelReservation(CancelProductReservation cmd)
        {
            var productId = cmd.AggregateId;
            var quantity = cmd.Quantity;

            if (quantity > Reserved) quantity = Reserved;
            RaiseEvent(new ProductReservationCancelled(productId, quantity));
        }

        internal void Reserve(ReserveProduct cmd)
        {
            var productId = cmd.AggregateId;
            var quantity = cmd.Quantity;

            if (AvailableQuantity < quantity)
            {
                RaiseEvent(new ProductReservationFailed(productId, quantity, ProductReservationFailed.NotAvailable));
            }
            else
            {
                RaiseEvent(new ProductReserved(productId, quantity));
            }
        }

        #endregion

        #region Helpers

        private void ChangeQuantity(int newQuantity)
        {
            var productId = AggregateId;
            var difference = newQuantity - Quantity;

            if (difference == 0) return;

            if (Quantity + difference < 0) throw new QuantityChangeException(AggregateId, $"Quantity can't be negative. Quantity: {Quantity}, Diff: {difference}");

            RaiseEvent(new ProductQuantityChanged(productId, difference));
        }

        #endregion
    }
}

File: PinetreeShop/PinetreeShop.Domain.Products/ProductAggregate.fs
module PinetreeShop.Domain.Products.ProductAggregate

open PinetreeCQRS.Infrastructure.Types
open PinetreeCQRS.Infrastructure.Commands
open PinetreeCQRS.Infrastructure.Events
open System
open Chessie.ErrorHandling
open PinetreeCQRS.Infrastructure.Validation

type ProductError =
    | ValidationError of string
    interface IError
    override e.ToString() = sprintf "%A" e

type Command = 
    | Create of string * decimal
    | AddToStock of int
    | RemoveFromStock of int
    | Reserve of int
    | CancelReservation of int
    | PurchaseReserved of int

type Event = 
    | ProductCreated of string * decimal
    | ProductQuantityChanged of int
    | ProductReserved of int
    | ProductReservationCanceled of int
    | ProductReservationFailed of int * IError list
    | ProductPurchased of int
    interface IEvent

module private Handlers = 
    type State = 
        { Created : bool
          Quantity : int
          Reserved : int }
        static member Zero = 
            { Created = false
              Quantity = 0
              Reserved = 0 }


    let applyEvent state event = 
        match event with
        | ProductCreated(name, price) -> { state with Created = true }
        | ProductQuantityChanged diff -> { state with Quantity = state.Quantity + diff }
        | ProductReserved qty -> { state with Reserved = state.Reserved + qty }
        | ProductReservationCanceled qty -> { state with Reserved = state.Reserved - qty }
        | ProductReservationFailed (qty, reasons) -> state
        | ProductPurchased qty -> 
            { state with Reserved = state.Reserved - qty
                         Quantity = state.Quantity - qty }

    module private Validate =
        let inCase predicate error value = 
            let result = predicate value
            match result with
            | false -> ok value
            | true -> Bad [ ValidationError error :> IError ]

        module private Helpers =
            let available state = state.Quantity - state.Reserved
            let notCreated = inCase (fun s' -> s'.Created) "Product already created"
            let positiveQuantity = inCase (fun q' -> q' <= 0) "Quantity must be a positive number"
            let positivePrice = inCase (fun p' -> p' < 0m) "Price must be a positive number"
            let canChangeQuantity = inCase (fun (s', d') -> not (d' < 0 && available s' >= -d' || d' >= 0)) "Not enough available items"
            let enoughReservedItems = inCase (fun (s', q') -> s'.Reserved < q') "Not enough reserved items"
            let created = inCase (fun s' -> not s'.Created) "Product must be created" 

        let canCreate (s, p) = Helpers.notCreated s <* Helpers.positivePrice p
        let createdAndPositiveQuantity (s, q) = Helpers.created s <* Helpers.positiveQuantity q 
        let createdAndEnoughReservedItems (s, q) = Helpers.created s <* Helpers.positiveQuantity q <* Helpers.enoughReservedItems (s, q)
        let createdAndCanRemoveItems (s, q) = Helpers.created s <* Helpers.positiveQuantity q <* Helpers.canChangeQuantity (s, -q)


    let executeCommand (state:State) command = 
        match command with
        | Create(name, price) -> Validate.canCreate (state, price) <?> [ ProductCreated(name, price) ]
        | AddToStock qty -> Validate.createdAndPositiveQuantity (state, qty) <?> [ ProductQuantityChanged(qty) ]
        | RemoveFromStock qty -> Validate.createdAndCanRemoveItems (state, qty) <?> [ ProductQuantityChanged(-qty) ]
        | CancelReservation qty -> Validate.createdAndEnoughReservedItems (state, qty) <?> [ ProductReservationCanceled(qty) ]
        | PurchaseReserved qty ->  Validate.createdAndEnoughReservedItems (state, qty) <?> [ ProductPurchased(qty) ]
        | Reserve qty -> 
            let r =  Validate.createdAndCanRemoveItems (state, qty) 
            match r with
            | Ok (s, _) -> ok [ProductReserved(qty)] 
            | Bad f -> ok [ProductReservationFailed(qty, f)] 


let makeProductCommandHandler = 
    makeCommandHandler { Zero = Handlers.State.Zero
                         ApplyEvent = Handlers.applyEvent
                         ExecuteCommand = Handlers.executeCommand } 